BSD Sockets API: Offloading Support

GAnthony opened the pull request:
Problem Statement:
=================
* See github issue #3706: (was Jira: ZEP-2271)

  "Users of devices which provide socket and TCP/IP offload engines
   would benefit in memory and power efficiency by enabling full
   offload of the Zephyr BSD socket APIs to a dedicated co-processor.

   The TI CC3220SF SoC, part of the CC32XX SimpleLink SoC family of WiFi
   enabled devices, will be used as the initial socket offload implementation."

* See Figure 1-2, http://www.ti.com/lit/ug/swru455c/swru455c.pdf,
  for the SimpleLink network co-processor architecture.

* In summary, it would be more efficient, in the case of vendors providing
  a complete BSD socket offload solution, to hook into the networking
  stack at the BSD socket API level rather than at the net_context() level
  (per the current Zephyr NET_OFFLOAD design);
  otherwise, for applications/protocols using the socket API:
  1) there is unnecessary overhead mapping Zephyr BSD sockets <-> net_context <->
     offloaded BSD socket APIs, converting between the synchronous socket
     and asynchronous callback-based net_context APIs;
  2) there is extra code and data brought in by instantiating the Zephyr
     networking stack (creation of a rx queue and thread, rx/tx pools)
     which is in principal already handled by the offload co-processor;
  3) to implement the mapping in 1), Zephyr network packets are required
     on the host MCU creating an extra buffer copy between the offload engine
     and the application's socket data buffers.

Proposed Solution:
=================
* Tap directly from the BSD socket API level, before net_context.

* Other Socket Offload Precedents: 
  1) Intel Socket Offload Framework: http://infiniband.sourceforge.net/archive/OSF_SDP_HLD.pdf
     This was designed for Linux, and was more general in that it allows multiple
     simultaneous socket providers in the system.

  2) MyNewt allows for "Socket Providers", enabling socket offload engines
     to register socket operations.  This RFC is more similar to MyNewt's solution:
     https://github.com/apache/mynewt-core/blob/develop/net/ip/mn_socket/include/mn_socket/mn_socket_ops.h

Validation:
===========
  1) http_sensor_demo: Demoed at Linaro Connect SFO17, a micropython script
     calling Zephyr socket APIs (via usocket), offloaded to a SimpleLink WiFi
     "socket provider", sending HTTP POSTs to dweet.io with onboard temperature
     sensor readings.
     https://github.com/GAnthony/micropython/blob/http_sensor_demo/ports/zephyr/scripts/http_sensor_demo.py          

  2) samples/net/sockets/echo: works unmodified given the
     prj_cc3220sf_launchxl.conf in this RFC.

Caveats:
========
* Only one socket offload provider allowed in the system at a time;
* As this bypasses the net_app/net_context APIs, existing protocols built on 
  net_app/net_context will not benefit from this socket offload;
  Note: There is now a plan to update all protocols to sockets: #7591.
* Though the SimpleLink provider also offloads DNS and DHCP, there is
  currently no way to offload that from Zephyr with this solution;
* This offload method is at a level which would bypass existing Zephyr IP
  routing between network interfaces;
* ~~WiFi Provisioning: this socket provider starts in Station mode by default,
  and the AP security key is passed in on the build command line.
  In the future, it could implement the various provisioning options
  supported by the chip, and be configured via Kconfig;~~
  A default provisioning using Fast Connect policy has been implemented.
* ~~The SimpleLink socket API supports select(), but is not yet exported by this
  provider, as Zephyr does not yet support select().~~
  poll() has been implemented over select().

Commit:
driver: wifi: simplelink: Register socket offload operations.

If the SimpleLink WiFi driver is configured, and socket offload
enabled, this revectors the Zephyr BSD socket APIs to the SimpleLink
WiFi host driver BSD socket APIs, providing a
direct offload of the TCP/IP stack to the CC3220SF network
coprocessor.

Fixes #3706

Signed-off-by: Gil Pitney <gil.pitney@linaro.org>

Commit:
net: sockets: Add a socket offload module

This patch enables BSD socket offload to a dedicated
TCP/IP offload engine.

This provides a simpler, more direct mechanism than going
through NET_OFFLOAD (zsock -> net_context -> socket conversions)
for those devices which provide complete TCP/IP offload at the
BSD socket level, and whose use cases do not require
IP routing between multiple network interfaces.

To use, configure CONFIG_NET_SOCKETS_OFFLOAD=y, and register
socket_offload_ops with this module.

Fixes #3706

Signed-off-by: Gil Pitney <gil.pitney@linaro.org>

Commit:
samples: net: sockets: echo: enable socket offload for cc3220sf_launchxl

Add a prj conf file for the TI cc3220sf_launchxl board
to enable socket offload to the simplelink WiFi driver.

Signed-off-by: Gil Pitney <gil.pitney@linaro.org>

jukkar reviewed (dismissed):
> As this bypasses the net_app/net_context APIs, existing protocols built on
> net_app/net_context will not benefit from this socket offload;

This is the most problematic issue with this solution. All the protocols we have built on top of native IP stack would be rendered useless from user point of view. This also fragments the IP stack and leads to applications that need to implement things itself instead of using the networking APIs provided by native stack.

We already have offloading API that is run under net_context, if that is used then all the current protocols and also BSD socket API will be offloaded. As the net_context APIs is modelled after BSD socket API (==it provides similar functions) and thus mapping BSD sockets over net_context API is quite straightforward.

Your time would be better spent if you could send patches that enhance/fix the current offloading solution instead of this proposal. IMHO we do not need this BSD socket offloading at all.

pfalcon commented on include/socket_offload_ops.h:
```diff
@@ -0,0 +1,56 @@
+ * It is assumed that these offload functions (except for init()) follow the
+ * POSIX socket API standard for arguments, return values and setting of errno.
+ */
+struct socket_offload {
```
Why this structure goes into a separate header, why we can't have single socket_offload.h?

GAnthony replied:
"Principle of minimum inclusion".  The idea is to only include what's needed.  The socket provider only needs the definitions in socket_offload_ops.h, but not the definitions in socket_offload.h.

pfalcon commented on subsys/socket_offload/socket_offload.c:
```diff
@@ -0,0 +1,35 @@
+/*
```
I'm almost sure all these files should go into subsys/net/lib/sockets/.

GAnthony replied:
Yes.  This patch was quite removed from the networking stack, I didn't want to presume.  But, yes, ideally.

pfalcon commented on include/socket_offload_ops.h:
```diff
@@ -0,0 +1,56 @@
+			  const void *optval, socklen_t optlen);
+	int (*getsockopt)(int sock, int level, int optname, void *optval,
+			  socklen_t *optlen);
+	ssize_t (*recv)(int sock, void *buf, size_t max_len, int flags);
```
In POSIX, recv(...) == recfrom(..., NULL). I think we should rely on that property, and have only recvfrom/sendto virtual methods.

GAnthony replied:
Sounds logical.  I imagine a similar simplification can be done for send():  i.e., send() == sendto(..., NULL, 0).
Then, some simplification also for zsock_ and net_context_ APIs as well?

pfalcon commented on subsys/socket_offload/providers/Kconfig.simplelink:
```diff
@@ -0,0 +1,31 @@
+# Kconfig.simplelink - SimpleLink Socket Offload Provider configuration
```
I'd say this should be treated as "socket offload driver", and thus live somewhere in drivers/

pfalcon commented on subsys/socket_offload/providers/simplelink.c:
```diff
@@ -0,0 +1,457 @@
+
+/* Excerpted from SimpleLink's socket.h:
+ * "Unsupported: these are only placeholders to not break BSD code.
+ *  Remove once Zephyr has POSIX socket options defined."
```
I guess sentence about Zephyr is not part of the quote from SimpleLink's socket.h.

GAnthony replied:
Yup, thanks.

pfalcon commented on subsys/socket_offload/providers/simplelink_support.h:
```diff
@@ -0,0 +1,27 @@
+#ifdef NET_SSID
+#define SSID_NAME       STRINGIFY(NET_SSID)
+#else
+#error "Pass CFLAGS="-DNET_SSID="<SSID_Name>" -DNET_PASS="<password>"" to make"
```
Shouldn't these be Kconfig options?

pfalcon commented on include/net/socket.h:
```diff
@@ -53,6 +53,9 @@ struct zsock_addrinfo {
 };
 
+/* Note: placing zsock_ symbols in <net/zsock.h>, and making zsock
+ * the default socket provider would greatly simplify this header.
```
Uo remind, Zephyr's socket API functions have prefix "zsock_". Then there's a separate aliasing layer for raw POSIX names. Someone offloading Zephyr socket API would naturally do that on the level of "zsock_" functions. I don't insist though (I may imagine doing that way leads to a shorter patch), just a notice that it may need to be redone later.

pfalcon reviewed (dismissed):
Some technically oriented comments.

pfalcon commented:
> We already have offloading API that is run under net_context, if that is used then all the current protocols and also BSD socket API will be offloaded. As the net_context APIs is modelled after BSD socket API (==it provides similar functions) and thus mapping BSD sockets over net_context API is quite straightforward.

My personal IMHO is that we don't need even that offloading API, and instead should "motivate" vendors to provide L2 integration. But the reality is different, where there's a need to motivate vendors use Zephyr at all (or they'll stick with other RTOSes).

Existing offloading API offloads everything except buffer management. And the motivation for socket-level offload was to offload that too, literally, if there's a dedicated engine with its own memory to handle networking data, then there's no need to spend application CPU's memory and cycles on managing the buffers.

The expectation that it would be a part of socket API was there right from the beginning, and came natural even for external observers, e.g. it was one of the first comments from a MyNewt developer with a proposal to standardize on the same socket offload vtable (i.e. MyNewt follows the same design (except maybe what we call "offload", for them is just driver (perhaps TCP/IP stack "driver") interface).

So, my motivation with this patch is simple - this was a subtask of BSD Sockets API "epic" from the beginning, and it's not a big, and pretty clean patch, which addresses requirements of some of the stakeholders (e.g. TI's), i.e. allows easier adoption of Zephyr.

tbursztyka commented:
> allows easier adoption of Zephyr.

And bring 2 different models of doing network in Zephyr, more maintenance, no compatibility/portability with native stack, etc etc...  How's that going to work on the long run once people will adopt Zephyr? I can foresee this already: it will become the next FreeRTOS...

So no.

If there are issues with current native stack and the way to optimize the offloading in it, let's do so. @jukkar comment summed things up pretty well imo.

pfalcon commented:
> This also fragments the IP stack

There's a definitely a "risk" of that, but the "problem" is not in this patch, but in BSD Sockets API. It's just some potential Zephyr users don't seem to value all the careful design of Zephyr IP stack and innovative solutions like data fragment based packet management, they just want the plain old sockets, because that's what they used all the time, and for what they have existing solution.

It's definitely a question where that leads us, but if Zephyr strives to be a general-purpose OS, it shouldn't be a surprise if people want to use it as such, including relying on well-known APIs.

pfalcon commented:
> And bring 2 different models of doing network in Zephyr, more maintenance, no compatibility/portability with native stack

Even with existing NET_OFFLOAD, compatibility/portability is heavily compromised, but e.g. you don't care that much ;-) : https://github.com/zephyrproject-rtos/zephyr/pull/4711#issuecomment-342083689

So, why be surprised if some people care even less of native IP stack (but apparently still value other part of Zephyr)?

> etc etc... How's that going to work on the long run once people will adopt Zephyr? I can foresee this already: it will become the next FreeRTOS...

Fairly speaking, that wouldn't be too bad, and even that would require quite a long run, with attention to users at each mile... (Of course doing  it "right" would be even better, e.g. if it were my project, I'd reject NET_OFFLOAD, lol).

tbursztyka commented:
> Even with existing NET_OFFLOAD, compatibility/portability is heavily compromised, but e.g. you don't 
> care that much ;-) : #4711 (comment)

Where did you see that I don't care? Again assuming things about what people think. wth?

And where do you see net_offload breaking things? Any app done on top of native stack, will run the exact same way on top of a net_offload device: it's the same apis!

pfalcon commented:
> Where did you see that I don't care? Again assuming things about what people think. wth?

Correction: you don't find that to be a big issue. That was an illustration that there's a pluralism of opinions, some think that drivers should integrate at the proper L2 layer, some think that can be lifted and integration can be above that, but still tied into Zephyr's buffer management scheme, and some think that buffer management can be "offloaded" too. So, "assume" is not the right word here, but we definitely should try to understand what other parties think, and what are patterns in that thinking, do understand the big picture.

> And where do you see net_offload breaking things? Any app done on top of native stack, will run the exact same way on top of a net_offload device: it's the same apis!

What I mean is that net_offload already bypasses a big part of Zephyr networking stack. Let me give an example: it's possible to have a very robust Zephyr networking stack (which it's not currently), and still have a Zephyr application (product) to be security-vulnerable, because it using 3rd party network context offloading feature, with vulnerability in that 3rd-party code.

> it's the same apis!

Yes, it's the same APIs, taking Zephyr's adhoc API as a baseline. As mentioned above, there're parties who apparently take BSD Sockets API as the baseline.

You and Jukka are absolutely right that it raises concerns about the protocol implementations done on the native API. But nothing in *this* patch has very specific conflicts and contradictions. When someone submits e.g. a libcoap port to be included in Zephyr tree, that would be the right time to raise big concerns of unneeded duplication. (Otherwise, it would be hard to preclude people from doing such a port, reusing existing code is the whole idea of the standard APIs.)

tbursztyka commented:
> Correction: you don't find that to be a big issue. That was an illustration that there's a pluralism of opinions, some think that drivers should integrate at the proper L2 layer, some think that can be lifted and integration can be above that, but still tied into Zephyr's buffer management scheme, and some think that buffer management can be "offloaded" too. So, "assume" is not the right word here, but we definitely should try to understand what other parties think, and what are patterns in that thinking, to understand the big picture.

In case of winc1500, if you use the Ethernet L2, then you loose the other features it can offload, afaik. The point of net_offload is to be able to use as much as possible the offloaded features. If it can do dhcpv4, why using ethernet mode which would need to bring up ethernet l2 and our dhcpv4 client into the build? And you still need the wifi management interface, either way.

About the buffer management, unless you can _transparently_ offload that, fine. But until then there will be this kind of copies. Actually, good luck to make this possible, because you still need to be able to route net_pkt throughout different bearers etc... (and I am not putting the user API in the picture here, just talking about how the core would have to manage this. It's not going to be easy).

All these problems, we've known them from day 1. This current PR looks like a very narrow look at how network offloading could be done. I fully understand that stakeholders may try to push for their framework when it's about their own unique comms chip, but within Zephyr design, it won't fit all the other use-cases. It is not a proper integration.

NET_OFFLOAD is far from being perfect, but from the start the way offloaded chip work make it hard to get it perfect (offloading high level APIs...)

pfalcon commented:
@tbursztyka : So, to clarify re: winc1500, I'm of course not saying that *you* could do it differently, what you do is taking somebody's stale PR and try to salvage it, which is already pretty great. But it could be done differently in general, we're just experimenting with doing it via NET_OFFLOAD. But Gil did just the same, he experimented with another way to offload, and well, it didn't turn out too bad (as long as we accept offload at all).

Anyway, I tried to do some "advocacy", to not leave Gil one on one with this, but will pass the word to him now. My point however is that this feature was always part of BSD Sockets work: https://github.com/zephyrproject-rtos/zephyr/issues/3706 , and now https://github.com/zephyrproject-rtos/zephyr/issues/3369 can't be closed until it's "done". And well, no concerns were raised about sockets offloading before, so it would be sad if it's done for nothing.

GAnthony commented:
Part of the reason for submitting this PR, if nothing else, was to highlight the dilemma faced by a vendor who provides a complete offloading solution, and considering to use Zephyr.

Far from being a perfect solution, it is attempting to resolve the issues stated in the problem statement:
1) Avoiding the overhead mapping Zephyr BSD sockets <-> net_context <->
    offloaded BSD socket APIs (extra code)
2) Instantiation of the rx queue and thread, and rx/tx pools  (extra data).
3) extra buffer copies between the offload engines buffers and the application's socket data buffers (lower performance).

Issue 2) may be ameliorated perhaps by judicious #ifdef CONFIG_NET_OFFLOAD sprinkled thoughout the net_core code, if that's acceptable, but I thought that would be rather intrusive.

I didn't see a way to to avoid issues 1) and 3) using NET_OFFLOAD.

So, that leaves a vendor evaluating Zephyr, who spent engineering effort to offload as much WiFi functionality as possible from the MCU, and (naturally) targeting a standard BSD sockets API, with a few choices:

a) Do full integration with the IP stack, at the L2 level.
    At this level, I'd expect IP routing, and everything else built on net_app to work.   But this does not give full hardware entitlement to the vendor's offload chip.

b) Tap off at the NET_OFFLOAD hook:   This  gives better hardware entitlement, but leaves issues 1) and 3) above unresolved.   This is mainly due to the mismatch in BSD socket API/linear buffer vs net_app/net_pkt buffers. 

c) Tap off at the BSD socket level:  This gives fuller hardware entitlement, and resolves the issues 1) and 3).  But, until/unless all protocols are written to the standard BSD socket APIs, the platform does not benefit from existing apps/protocols written to net_app/net_context.

d) Don't attempt to integrate with data plane of the Zephyr IP stack (data plane).  SimpleLink SDK is in fact designed to replace the local IP stack, as applications can #include <sys/socket.h> and just call the standard BSD socket APIs directly.  Just use Zephyr for sensor/actuator drivers and threading.

So, bottom line, are there any ways to resolve issues 1) and 3) with NET_OFFLOAD, while still providing the vendor full use of their offload hardware?

nashif commented:
There are too many topics of discussion here, the main point however is that when you introduce an offload api into Zephyr, it should not be fixed to work at one level only, as we have seen, there are different levels:
- L2
- Buffer
- Socket

There are use cases for all three and there is no good or bad offloading, it will all depend on the HW being used and how much some is willing to invest making their existing networking application work with Zephyr.

So the first question to answer, do we want to support offloading on all level? So basically what @tbursztyka  is doing with WINC, what @GAnthony is doing with SimpleLink and what @pfalcon mentioned re offloading on L2 layer with WINC. I would say yes, we need to look into a way for supporting all 3, if possible.

The main issue here is what happens if you have any of the 3 and how do you support existing protocols and middleware in Zephyr with any of the three cases above. The first 2 cases are probably covered right now due to the fact that protocols are implemented at buffer level, however, this PR expects offloading on the socket level and we do not have much support for sockets on any of the protocols, because they are implemented using native APIs.

To make this PR useful, you will need to implement everything on top of the IP stack using sockets duplicating existing code (or replacing it if you want to go all the way), so, a few questions:

1. Is the socket API ready for this?
2. What are the penalties that comes with that, for example HTTP using sockets vs using native
3. Can we make everything use sockets? Do we actually need to make everything use sockets?
4. Does offloading on socket level still require services that would go to native IP stack? DHCP? DNS? How is this handled?
5. Who is going to do all of that?

I personally have no issues having things like HTTP and MQTT, etc. using sockets instead of native stack to allow offloading on all level, but is this the end of it? I do not think so, the level is in the details. See questions above.

We wanted to dedicate next week's TSC meeting to this topic, so can someone please take the AI to structure all of this and present the problem we are trying to solve and a proposal for solving in the TSC next week?

nashif commented:
recheck

jukkar commented:
In yesterdays TSC meeting, someone suggested that in order to support both socket offloading and current application protocol APIs (net-app), we could add BSD socket support into net-app. That would cause some extra memory usage and copying but might be acceptable. The socket support in net-app should be optional so that if one does not have socket offloading enabled, then existing native APIs would be used.
This would have the benefit that applications could either use BSD socket APIs directly, or if for example TLS needs to be supported, then net-app could be used too and that would still support socket offloading.

erwango commented:
> The socket support in net-app should be optional so that if one does not have socket offloading enabled, then existing native APIs would be used.
This would have the benefit that applications could either use BSD socket APIs directly, or if for example TLS needs to be supported, then net-app could be used too and that would still support socket offloading.

So, if socket offloading is used, both net_app and BSD socket API would be available?
No restriction at all?

nashif commented:
> The socket support in net-app should be optional so that if one does not have socket offloading enabled, then existing native APIs would be used.

why do you make this dependent on offloading, one could still use sockets without offloading

jukkar commented:
> why do you make this dependent on offloading, one could still use sockets without offloading

Sure, I did not say anything contrary. Socket API is available for applications as is. What I meant that if there is no socket offloading, then net-app would use net-context API directly instead of using socket API.

pfl commented on include/socket_offload_ops.h:
```diff
@@ -0,0 +1,56 @@
+			  const struct sockaddr *to, socklen_t tolen);
+};
+
+extern void socket_offload_register(const struct socket_offload *ops);
```
Hmm, all the functions in socket_offload_ops.h have identical functionality in include/net/net_offload.h. Was there any reason include/net/net_offload.h could not be (re)used here or modified to fit this use case?

GAnthony replied:
Well, those functions require a net_context structure, and have callback arguments; and though they have similar names, the behaviour and usage is quite different.  The goal here is to provide an interface to BSD socket APIs provided by a vendors TCP/IP offload engine.

jukkar commented:
After pondering this socket offloading a bit more and making a small prototype where net_app would use BSD socket API, I found out following:
* This could be make to work but it would probably make not much sense. Reason being that then all the calls to net_context would not go anywhere. So if either our internal APIs or applications would call net_context, then those calls would fail in this scenario.
* If we then go and make net_context to call BSD socket, we would end up in the same situation what we have now, the current net_offload API.

So it would make lot of sense from API point of view, that offloading is only done under net_context which is the current proposed solution for offloading. This allows current implementations of net_app, net_context and BSD socket to work as is without introducing any API fragmentation.

nashif commented:
> This could be make to work but it would probably make not much sense. Reason being that then all the calls to net_context would not go anywhere. So if either our internal APIs or applications would call net_context, then those calls would fail in this scenario.

If you are offloading on the socket level, why would anything go to net_context? AFAIK the socket is then implemented in externally to Zephyr and any socket calls would be offloaded and would not go down the zephyr IP stack.

pfl commented:
> If you are offloading on the socket level, why would anything go to net_context? AFAIK the socket is then implemented in externally to Zephyr and any socket calls would be offloaded and would not go down the zephyr IP stack.

Because the rest of the applications created so far also should be working also on this hardware, once it is supported by Zephyr? Like internal DNS, CoAP and all the other more or less weird network applications people have been using Zephyr for?

nashif commented:
> Because the rest of the applications created so far also should be working also on this hardware, once it is supported by Zephyr? Like internal DNS, CoAP and all the other more or less weird network applications people have been using Zephyr for?

That is an issue we are aware of and it was discussed. Applications and protocols in Zephyr right now would not work, but that does not prevent us from supporting offloading at this level with the caveat that you have to bring your own applications using sockets. Still a valid use case. Zephyr is not a take it all or nothing at all.

tbursztyka commented:
@nashif : This offloading won't be valid for anything BUT 1 device. I really tried to explain that many times: it does not make any sense. Sure you'll get this cc3220 working with it... and? What if somebody wants to hook a cc2520 along with this chip? Done properly through NET_OFFLOAD it would be possible to have other bearers (with or without native drivers) working along and all being able to get packet routed through etc... All devices, offloaded or not MUST be accessed through the same API.

The BSD socket layer done by @pfalcon  would enable people using socket and that's it. 
Fact they won't be able to use net_app or all natively supported high level protocol is a side issue, that will need to be addressed in some other discussion.

tbursztyka reviewed (dismissed):
blocking this.

jukkar commented:
> That is an issue we are aware of and it was discussed. Applications and protocols in Zephyr right now would not work, but that does not prevent us from supporting offloading at this level with the caveat that you have to bring your own applications using sockets. Still a valid use case. Zephyr is not a take it all or nothing at all.

As I mentioned in my earlier comment, if we would offload under net_context these issues would go away. All the current APIs (net_context, net_app and BSD socket API) would continue to work just fine as is. I do not really understand this desire to offload at socket level.

pfalcon commented:
> I do not really understand this desire to offload at socket level.

I believe Gil explained it, including at the yesterday's meeting: using net_context requires dealing with net_pkt/net_buf's in the main memory (spending main memory on net_pkt/net_buf's, converting data user data <-> net_pkt/net_buf <-> offload engine data (usually socket user data)).

Nothing that is really needed (well, it wastes user's memory for nothing) if there's a dedicated offload engine with dedicated memory space.

Such designs are many, sockets is *the* standard networking interface, and many WiFi devices offer that as the black box.

pfalcon commented:
> This offloading won't be valid for anything BUT 1 device.

Absolutely. When I looked at Zephyr's BLE API, I found that functions don't take a BLE adapter as a param. Asking @jhedberg, the response was along the lines of "well, we target minimum resource usage". World's not perfect.

So yeah, let the users of the socket offloading API suffer being able to use only 1 device.

nashif commented:
> This offloading won't be valid for anything BUT 1 device.

not entirely true. Many devices and Wifi chips support this mode. You can have Wifi on an Arduino Uno, without having an IP stack. AFAIK the WINC1500 supports this as well.

tbursztyka commented:
> not entirely true. Many devices and Wifi chips support this mode. You can have Wifi on an Arduino Uno, without having an IP stack. AFAIK the WINC1500 supports this as well.

And? That's the same problem. And btw, winc1500 is going to be enabled through NET_OFFLOAD, not this socket/ti-use-case stuff.

GAnthony commented:
The Problem Statement at the top of this thread motivates the reasons for this alternate method to NET_OFFLOAD, as well as the known issues with this PR, which have been basically restated throughout the conversation.

One goal of the SimpleLink SDK is to essentially replace the local IP stack of the host OS, and so it provides many services like BSD sockets, DNS, DHCP, HTTP server, and even some higher level protocols (built on sockets).   The typical use case has only the WiFi connection from an edge IoT device, transmitting data/commands to an access point.

So, the alternative to NET_OFFLOAD (tapping off at net_context level) or SOCKET_OFFLOAD (tapping off at socket level), is to directly call the APIs of the SimpleLink host driver.  This would be the most efficient, avoiding extra function calls, or the socket buffer to net_pkt copy overhead of the NET_OFFLOAD solution.
See PR #5028, board/samples/cc3220sf_launchxl/echo sample.
 
I think it's reasonable to show customers a way to directly use the product's SDK, using BSD socket APIs, as long as they realize it's not deeply integrated with the Zephyr IP stack, meaning:
* multiple bearer support would bypass the SimpleLink WiFi interface;
* any protocol libraries they want working over WiFi need to be available/ported to sockets;
* until the control plane APIs are available in Zephyr, they would have to call vendor-specific WLAN and network configuration APIs (or rely on configurable default methods).

Customers still use Zephyr APIs for their sensors and actuators, multithreading, and other apps.

So, given that:
* this conversation is yielding no consensus; and
* PR#5028 at least shows a means for TI customers to use WiFi in Zephyr now; 

I propose shelving this PR, to perhaps resurrect, when/if there becomes some critical mass of Zephyr protocols which actually get built on sockets.

GAnthony commented:
Given,
a) There is now a plan to move Zephyr protocols to sockets (#7591), removing the previous main objection to this PR; and
b) there is potential interest from another Si vendor (https://github.com/zephyrproject-rtos/zephyr/pull/8814#discussion_r201637136),

This PR is being reopened to continue to the discussion, hopefully to a resolution agreeable by all parties.

The changes from the previous PR, taking into account some of the earlier feedback, include:
* Moved socket offload module into subsys/net/lib/sockets;
* Removed the idea of a separate "socket provider" entity, and moved the SimpleLink socket offload functions and registration into the SimpleLink WiFi driver, to keep both data and control plane offload operations in the same directory;
* Renamed CONFIG_SOCKET_OFFLOAD to CONFIG_NET_SOCKETS_OFFLOAD;
* For the echo example, enabled the new net wifi shell, so users can connect to an AP with SSID and password before sending data to the echo server.  This avoids the need to build a secret password into the image.  This is in lieu of more advanced WiFi provisioning to be added in the future.

As TLS support is currently being added to the Zephyr socket layer, and the goal of this socket offloading is to also offload *secure socket* operations (cyrpto, TLS handshaking) to the TOE hardware, some better integration with that work may be needed in the future (than the simple #ifdef switch provided here).

dbkinder commented on samples/net/sockets/echo/README.rst:
```diff
@@ -76,3 +76,22 @@ To test:
+    wifi> scan
+    wifi> connect "<mySSID>" <SSID_len> 0 <myPassword>
+
+replacing <*> tokens with the actual SSID name, length, and password.
```
change to

     ``<*>``

GAnthony replied:
Thanks.
Actually, when the patch is re-based and "fast connect" feature enabled, these instructions get simplified.  The mention to ''<*>" is removed.  Next push will have an updated section on cc3220sf_launchxl.

dbkinder reviewed (dismissed):
one nit fix

jukkar commented on samples/net/sockets/echo/prj_cc3220sf_launchxl.conf:
```diff
@@ -0,0 +1,22 @@
+# Networking Config:
```
We are trying to migrate away from having a new "full" conf file for each board. Please see for example how samples/net/echo_server is doing it and place board specific options into boards/ directory.

GAnthony replied:
Thanks for the review.

Yes will move conf file into /boards.

jukkar commented on include/net/socket.h:
```diff
@@ -144,6 +144,7 @@ int ztls_setsockopt(int sock, int level, int optname,
 #endif /* defined(CONFIG_NET_SOCKETS_SOCKOPT_TLS) */
```
Please change the commit subject to be something like "net: sockets: offload: ...."

GAnthony replied:
OK

jukkar commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -62,6 +62,12 @@ config NET_SOCKETS_TLS_MAX_CIPHERSUITES
 	  available to the socket.
 
+config NET_SOCKETS_OFFLOAD
+	bool "Offload Socket APIs"
```
Add [EXPERIMENTAL] at the end of the description

GAnthony replied:
Agreed.

jukkar commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -62,6 +62,12 @@ config NET_SOCKETS_TLS_MAX_CIPHERSUITES
+	bool "Offload Socket APIs"
+	select NET_SOCKETS_POSIX_NAMES
+	help
+	Enables direct offloading of socket operations to dedicated TOE hardware
```
Also add information about restrictions when this offloading should not be used (multiple network interface with routing etc).
BTW, what does "TOE" here means, please open this up.

GAnthony replied:
Yes, OK.
TOE is a general term, which I think applies to this case:  
https://en.wikipedia.org/wiki/TCP_offload_engine

jukkar commented on samples/net/sockets/echo/README.rst:
```diff
@@ -76,3 +76,15 @@ To test:
 
```
Usually for samples, the commit subject starts with "samples: net: ...."

GAnthony replied:
OK

rlubos commented on include/net/socket_offload_ops.h:
```diff
@@ -0,0 +1,54 @@
+ * It is assumed that these offload functions follow the
+ * POSIX socket API standard for arguments, return values and setting of errno.
+ */
+struct socket_offload {
```
Is there a plan to have any offload-specific fields in this structure? Because at this point it seems to cover socket API, so perhaps we could call it `struct socket_api`?

I'm asking, because I have an image in my mind for improving current POSIX sockets/secure sockets coexitstence. Instead of having separate `#ifdef`s  in every socket API function, we could have wrappers like in `socket_offload.h` and separate set of of API structures for POSIX sockets, secure sockets and whatever offloaded sockets we want. Then we could simply switch between them during compilation (or even runtime in the future?).

rlubos replied:
BTW, why is there no `poll`?

GAnthony replied:
> BTW, why is there no poll?

Well, SimpleLink offload only supports select(), and I'd rather have select() than poll(), as it seems to me that more socket based applications and tutorials expect select().
I agree that one or the other should be added (ideally both), and a "not implemented" error returned if the offload chip doesn't support one or the other (and, in future, could implement poll over select or vice-versa).

GAnthony replied:
Thanks for the feedback.

> Instead of having separate #ifdefs in every socket API function, we could have wrappers like in socket_offload.h and separate set of of API structures for POSIX sockets, secure sockets and whatever offloaded sockets we want. Then we could simply switch between them during compilation (or even runtime in the future?).

A good thought.   
This PR adds yet another #ifdef to socket.h, which is starting to get messy.

My original thought was that Zephyr zsock could be the default "socket provider", and both variants of (zsock, offload) can be registered via something like `socket_register(const struct socket_ops *ops)`.

Some capabilities to consider:
1) The SimpleLink offload driver handles secure sockets through the same set of APIs as non-secure (just different flags);
2) It is possible for a non-secure socket to be "upgraded" to a secure socket via a setsockopt() call. 
3) It is possible to mix secure and non secure client threads together in the same application.

With the global #ifdef CONFIG_NET_SOCKETS_SOCKOPT_TLS , type compile switches, I guess points 2) and 3) aren't yet possible.    Maybe a runtime switch could help.

If you are close to maybe posting a WIP PR with this `struct socket_api` concept, perhaps we can merge the two?

rlubos replied:
> I agree that one or the other should be added (ideally both)

I think this is the right way, have both specified in the API, and leave up to implementation, which ones are implemented. I just wanted to point it out, that existing socket samples in Zephyr uses `poll`, as it is currently available there, so if we wanted to use samples out-of-the-box (or with minor modifications) with offloaded driver, we'd have to have poll implemented.

> Some capabilities to consider:
> The SimpleLink offload driver handles secure sockets through the same set of APIs as non-secure (just different flags);
> It is possible to mix secure and non secure client threads together in the same application.

That's the same (or similar) as we have in Zephyr now. Secure sockets API proposed recently can handle both, secure protocols and non secure. So you can use the same API to create regular TCP/UDP sockets and secure TLS/DTLS sockets in the same application. The decision which one should be created is made based on protocol type provided during socket creation. In case TCP/UDP is used, socket API calls are simply forwarded to the `zsock` implementation. There is no example showing this functionality yet, but it is possible.

> It is possible for a non-secure socket to be "upgraded" to a secure socket via a setsockopt() call.

That's indeed impossible with current implmentation.

> With the global #ifdef CONFIG_NET_SOCKETS_SOCKOPT_TLS , type compile switches, I guess points 2) and 3) aren't yet possible. Maybe a runtime switch could help.

I hope I've already answered this. Let me know is something is still unclear.

> If you are close to maybe posting a WIP PR with this struct socket_api concept, perhaps we can merge the two?

Well, it was just general idea and I dont' have any code yet, and I'm afraid I can't do it now. I'm going on a vacation in a week and I have other work to finish before that. But as I said, I view this PR as a good base for future improvement, and this can be one of them :)

rlubos reviewed (commented):
Thank you for restoring this PR!

Looks like good starting point for TLS offloading and further improvments.

I'm not approving yet as there are comments from others to be refered to. Also left one comment to consider.

GAnthony commented:
Rebased, updated per @jukkar comments.
Still need to address poll/select API.

GAnthony commented:
Rebased, and also addressed request for poll() by @rlubos.

* implemented poll() API over SimpleLink's select() API.   Chose not to export select() at this time, until/unless Zephyr supports it;
* Validated with socket echo_client sample, both UDP and TCP;
* Also, added logic to handle the MSG_DONTWAIT recv() flag.

GAnthony commented:
reheck

GAnthony commented:
recheck

rlubos reviewed (approved):
Thank you for the update. Good to see that there's also an example demonstrating socket offload.  :+1:

pfalcon commented:
Sorry folks, I wasn't able to follow thru recent changes/discussion here, but regarding poll/select stuff: there's no other way but to move to epoll(), as poll() doesn't scale (and select() doesn't scale at all, though depends on what to call "scale" of course). So, please keep that in mind. I just posted a mail titled "[RFC] Thinking about extended poll support in Zephyr", feel free to skim thru it and raise any concerns you see even on the entrance to it.

GAnthony commented:
@pfalcon, thanks for the note on epoll:

from: https://en.wikipedia.org/wiki/Epoll
>  It is meant to replace the older POSIX select(2) and poll(2) system calls, to achieve better performance in more demanding applications, where the number of watched file descriptors is large ...

I wonder if our target IoT devices will have such a large number of simultaneous open sockets, which would require such an optimization?

pfalcon commented:
> I wonder if our target IoT devices will have such a large number of simultaneous open sockets, which would require such an optimization?

Well, I guess it would be better to discuss this matter in the thread on the devel mailing list, the message which I posted there is at https://lists.zephyrproject.org/g/devel/topic/rfc_thinking_about_extended/25004178 .

Answering your question, everything is relative. For example, I think that "our target IoT devices" don't require such complications as user/kernel mode split, and yet it's in Zephyr. Nor I think that they require PTP, 802.1Qav protocols, and yet they're there too.

Regarding epoll, my motivation is:

1. epoll was in the original stakeholders' docs for implementing BSD Sockets API. It's not implemented, so I still consider it on my plate.
2. In the mail above, I give 2 examples, when poll()-like interface requires marshalling pollfd-like structure back and forth. That's not good imho.

And how much is "a large number of simultaneous open sockets"? I don't think that 1000 connections is a large number, even for IoT. But that's already quite a large number of poll descriptors to shove back and forth on each call.

pfalcon commented:
Not that everyone should like epoll, because it's a natural optimization of translating different polling mechanisms between themselves. For example, if you underlying mech is select(), then to emulate poll(), you need on each poll() call create a new fdset, and set all the bits around. With epoll, your internal representation is cached: you create fdset ones on epoll_create(), patch bits on epoll_ctl(), translate back just a little bit of it on epoll_wait().

jukkar reviewed (approved):
LGTM. The epoll() support can be added later so we can start with this atm.

jukkar commented:
There is the RFC status marked to this PR so I am wondering is this ready to be merged?

GAnthony commented:
> There is the RFC status marked to this PR so I am wondering is this ready to be merged?
@jukkar, I just removed the RFC status.

xiaodongxusc commented:
Hi,

I'm currently evaluating BSD sockets offloading on TI cc3220sf_launchxl. I'm working on zephyr master branch with all 3 commits for this case has been merged. However I got a build error
simplelink_sockets.c:117:25: error: ‘SlInAddr_t {aka struct SlInAddr_t}’ has no member named ‘S_un’
     sl_addr_in->sin_addr.S_un.S_addr;

It looks like sl_socket.h is not updated with adding S_un to 
typedef struct SlSockAddrIn_t
{
    _u16              sin_family;         /* Internet Protocol (AF_INET).                    */
    _u16              sin_port;           /* Address port (16 bits).                         */
    SlInAddr_t        sin_addr;           /* Internet address (32 bits).                     */
    _i8               sin_zero[8];        /* Not used.                                       */
}SlSockAddrIn_t;

My question is
1. What's the current status of this BSD socket offloading feature?
2. When do you think this feature will be completed&tested on master branch?

Thanks,
Xiaodong

GAnthony commented:
>  However I got a build error
simplelink_sockets.c:117:25: error: ‘SlInAddr_t {aka struct SlInAddr_t}’ has no member named ‘S_un’
sl_addr_in->sin_addr.S_un.S_addr;

@xiaodongxusc, Thanks for evaluating.
Please ensure to update to latest master, and try to build the sockets/echo sample. 
This build error should have been fixed last week, with the commit
be64964f1ec36eab84f3c849b6824bd6da106190 "driver: wifi: simplelink: Fix socket offload after s_addr cleanup"

GAnthony commented:
> What's the current status of this BSD socket offloading feature?

As of last week, the feature worked for in-tree samples/net/socket/echo app.
Recently however, migration to the new Zephyr shell broke the samples/net/wifi app, which was required as a prerequisite for SimpleLink samples to provision the SSID/password for the WiFi Access Point.  This is in 
process of being debugged. #10617

Note that WiFi testing is not yet part of Zephyr CI loop, which suggests functionality breaks may occur in master.  However, build testing has at least been recently added, so that issue you just hit should not happen again.

> When do you think this feature will be completed&tested on master branch?

Here is list of TODO:
1) Support offline provisioning of certificates (specifying cert/key filenames) for the new TLS setsockopt() feature.
2) Support non-blocking sockets via fcntl() API - that will be needed by some networking apps and protocols.
3) Add sync point for applications to wait until a device connects to an AP, before starting communications.
4) Update to latest SimpleLink SDK - the one in Zephyr is quite old. 

Of course, there might be other features that would need to be added depending on use cases which might arise.  For example, more options for WiFi provisioning, or new offload hooks depending on needs of the new socket-based networking protocols.

oberstet commented:
rgd the problem statement of this issue

> "Users of devices which provide socket and TCP/IP offload engines would benefit in memory and power efficiency by enabling full offload of the Zephyr BSD socket APIs to a dedicated co-processor.

It doesn't mention the price one has to pay for this "feature":

* adding a big security attack surface which is in a closed blob (firmware on network coprocessor) and unproven (it will have bugs)
* under exclusive 3rd party control (TI), no way to inspect, "blind trust" model
* I am now running a proprietary niche stack. fantastic.
* not being able to use L3+ features of the Zephyr stack (eg cross interface IP routing)
* this problem only becomes worse with offloading not only TCP/UDP, but also TLS .. 
* etc .. should I continue?

For me, above is a no go. As a Zephyr user, I am looking for the exact opposite: remove everything possible outside of the OS. I _want_ to use all of Zephyr. I don't need a 2nd protocol stack. I need an RF frontend with MAC with (one) MCU core in one chip, and an L2 MAC level driver for Zephyr. Can the TI chip be used like this at all? If not, what's a recommended chip with Zephyr?

Rgd "offloading": the only offloading the RF/MAC frontend should do is L2 packet filtering and possibly VLAN stuff. The frontend should NOT have any kind of firmware of its own. Is there a (single) chip like this?

I should note that I am far from an embedded engineer (so there might be some fundamental flaws in my thinking), more like cloud etc and coming to this field with some strong opinions;)

Anyways, I would be interested in your opinions/perspectives/corrections! thanks

> btw: Zephyr rocks! so thanks for Zephyr in the first place;)

jukkar commented:
I agree with you here, the offloading stack is quite evil as it complicates the native stack design. The offloaded stacks were supported in order to get more boards to run zephyr. I am hoping that in the future different vendors would support their hw using native IP stack instead.
I think the original use case for these offloaded boards is to support Arduino which typically do not provide a host stack.

nandojve commented:
Hi @oberstet ,

I think interesting this topic, few questions:

> I am now running a proprietary niche stack. fantastic.
What is/are your radio?

What radio do you would like to see in Zephyr that runs in Linux (or any other open source stack) that meet your criteria?

Just curios, something like an HCI for WIFI, can it be solution?

oberstet commented:
thanks for the feedback, much appreciated! 

@jukkar :

>  .. in order to get more boards to run zephyr.

ok, got it. yes, "broader HW support" is definitely good! unless, IMO:

* it requires a big design/impl. burden in the core project? eg it should not add a price that all users (of all HW) have to pay (eg by introducing additional internal layers)
* it confuses users: what HW _should_ I use? what are the downsides/upsides? what is actually _recommended_? (see below)

> I am hoping that in the future different vendors would support their hw using native IP stack instead.

yes, that would be great!

fwiw, it took Linux years to gain gravity and nudge vendors to go OSS, and not only that, but mainline. some vendors (eg NXP/Freescale or Realtek) are actively following an upstreaming/mainline approach. and for me, this is one of the most important criteria to select HW.

however, I'm afraid, some vendors might follow a different path for (perceived) business reasons. as in: vendor lock in comes in nicely (for the vendor) with an offloaded stack. they want to drag in as much of "protocol stack" as possible. 

anyways, IMO, Zephyr should help users not only by providing the broadest HW support possible, but crucially also recommendations or clear advantages/disadvantages tables. eg the costs I listed in above for going with an offloaded network stack chip (single or external "modem chip").

but that's a politicial Q of couse: does Zephyr want to _encourage_ users to select HW that allows to follow a "good approach" (hosted stack, L2 MAC interface) rather than HW that _is_ supported by Zephyr, but technically and security wise inferior?

IOW: as a user coming to Zephyr I might want to know which chips are _recommended_ in a category (eg single chip incl. wifi). In my eyes, anything involving a proprietory blob and offloaded stack _cannot_ be in this list.

-----

@nandojve 

*In this context, I am looking for a single chip Arm Cortex-M33 microcontroller (256kB+ RAM, 1MB+ flash) with Wifi, where the Wifi part _only_ contains the RF frontend and L2 MAC. And supported by Zephyr of courses;) Does anyone know one?*

On Linux (not talking about here, but ..) I would choose Linux mac80211 (https://wireless.wiki.kernel.org/en/developers/Documentation/mac80211) supported HW, eg Atheros ath10k.

For me, I want to minimize or eliminate blobs as far as possible. In a pragmatic way.

fwiw, a non-pragmatic, but even more extreme approach: SDR and OSS

https://github.com/open-sdr/openwifi
https://github.com/open-sdr/openwifi-hw

in this case, even the HW is OSS=) note: this is not a practical approach sadly .. eg certification etc .. but

oberstet commented:
one more note rgd L2/MAC level interfacing from Zephyr for _wireless_ L2 technologies and regulatory compliance - here is what Linux does for 802.11 where portions of L2 code run in Linux kernel .. and hence compliance depends on that code: see the "Statement" here: https://wireless.wiki.kernel.org/en/developers/regulatory/statement - Also see:

* https://wireless.wiki.kernel.org/en/developers/regulatory
* https://wireless.wiki.kernel.org/en/vendors/vendorsupport
* https://wireless.wiki.kernel.org/en/developers/documentation/cfg80211

jukkar commented:
Having mac80211 supported in Zephyr would be very nice indeed. We are still quite far from it atm.
